File: 1-intro/4-exploring-the-waters/js/palindromeRearranging.js
const solution = str => {
  let aux = str.split('').sort();
  let count = 0;

  while (aux.length) {
    let c = aux.shift();

    if (c == aux[0]) {
      aux.shift();
    } else {
      count++;
    }
  }

  return (count <= 1);
}

/* -------------------------------------------------------------------------- */

const solution2 = str => str.split('').sort().join('').replace(/(\w)\1/g,'').length <= 1;

/*
A expressão regular /(\w)\1/g captura um caractere qualquer e verifica se existe
um outro igual do lado dele.
*/

File: 1-intro/4-exploring-the-waters/cpp/palindromeRearranging.cpp
bool solution(string str) {
    int arr[256] = { 0 };

    for (char c : str) {
        arr[c]++;
    }

    int count = 0;
    for (int i = 0; i < 256; i++) {
        count += arr[i] % 2;
    }
    
    return count <= 1;
}
